File: src/main/java/nl/joelchrist/shorty/shorties/managers/ShortiesManager.java
package nl.joelchrist.shorty.shorties.managers;

import nl.joelchrist.shorty.exceptions.DuplicateShortyException;
import nl.joelchrist.shorty.exceptions.EntityNotFoundException;
import nl.joelchrist.shorty.shorties.domain.Shorty;
import nl.joelchrist.shorty.shorties.repositories.ShortiesRepository;
import nl.joelchrist.shorty.util.ShortyUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.swing.text.html.parser.Entity;
import java.util.Optional;

@Service
public class ShortiesManager {
    @Autowired
    private ShortyUtil shortyUtil;

    @Autowired
    private ShortiesRepository shortiesRepository;

    public Shorty createShorty(String fullUrl) {
        String shortIdentifier;
        do {
            shortIdentifier = shortyUtil.generateShortIdentifier();
        } while (shortiesRepository.findShorty(shortIdentifier).isPresent());
        Shorty shorty = new Shorty(fullUrl, shortIdentifier);
        return createShorty(shorty);
    }

    public Shorty createShorty(Shorty shorty) {
        Optional<Shorty> shortyOptional = shortiesRepository.findShorty(shorty.getShortIdentifier());
        if (shortyOptional.isPresent()) {
            throw new DuplicateShortyException(shorty.getShortIdentifier());
        }
        shortiesRepository.saveShorty(shorty);
        return shorty;
    }

    public Shorty findShorty(String shortIdentifier) throws EntityNotFoundException {
        return shortiesRepository.findShorty(shortIdentifier).orElseThrow(() ->
                new EntityNotFoundException(Shorty.class, shortIdentifier));
    }
}

File: src/main/kotlin/nl/joelchrist/shorty/shorties/managers/ShortiesManager.kt
package nl.joelchrist.shorty.shorties.managers

import nl.joelchrist.shorty.exceptions.DuplicateShortyException
import nl.joelchrist.shorty.exceptions.EntityNotFoundException
import nl.joelchrist.shorty.shorties.domain.Shorty
import nl.joelchrist.shorty.shorties.repositories.ShortiesRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ShortiesManager(@Autowired private val shortiesRepository: ShortiesRepository) {
    fun create(entity: Shorty): Shorty =
            (entity.identifier.let { shortiesRepository.findByIdentifier(it) })
                    ?.let { throw DuplicateShortyException(entity.identifier) }
                    ?: entity.also(shortiesRepository::save)

    fun findByIdentifier(identifier: String): Shorty = shortiesRepository.findByIdentifier(identifier) ?: throw EntityNotFoundException(Shorty::class, identifier)

    fun findByOwner(owner: String): List<Shorty>? = shortiesRepository.findByOwner(owner)

    fun all(): List<Shorty>? = shortiesRepository.findAll()

    fun delete(identifier: String): Boolean = shortiesRepository.delete(identifier).takeIf { it == true } ?: throw EntityNotFoundException(Shorty::class, identifier)
}